tests/test_auto_parser.py

```diff
@@ -64,3 +64,20 @@ def test_auto_parser_with_fname_mixed(tmp_path):
     with pytest.raises(ValueError, match="duplicate key inserted: fname"):
         query = Q_AND(Q_EQ("col1", "4"), Q_EQ("fname", "col2"))  # to show confusion that can arise with mixed structure
         read_partitioned_table(f"file://{tmp_path}/", query, AutoParser(parse_filenames_as="fname"))
+
+
+def test_auto_parser_fname_with_from_str(tmp_path):
+    """Tests AutoParser works when combining filtering through 'from_str' method and fname query."""
+    partition1 = tmp_path / "col=a"
+    partition1.mkdir(parents=True)
+    df1.to_json(partition1 / "a1.json", orient="records", lines=True)
+    partition2 = tmp_path / "col=b"
+    partition2.mkdir(parents=True)
+    df2.to_json(partition2 / "a2.json", orient="records", lines=True)
+    df3.to_json(partition2 / "a1.json", orient="records", lines=True)
+
+    parser = AutoParser.from_str("col=b/fname", parse_filenames_as="fname")
+    case2_result = read_partitioned_table(f"file://{tmp_path}/", Q_EQ("fname", "a1.json"), parser)

```

Comment:
Check [the last commit](https://github.com/AmpX-AI/fsql/pull/35/commits/0ec42fd688908fc700d15e36809227164ed89b5a). Does it make sense to support this as well? It would allow anyone to combine filtering in the string (using `from_str` method) with filename queries. The only thing I do not like is the extended `is_terminal_level` method.

Reply:
yeah, good catch -- consistency is important, so we should support the generation/from_str. It does not look well one has to supply the parse_filenames_as argument -- I think we should derive it from that str as well -- it is there, and without the `=`. But that may be a risky business (I think I may have realized some time ago it is not possible due to some corner case).

I don't like the `is_terminal_level` either -- as well as the whole parser interface, as evidenced by the number of type-ignores and now also kwargs there. I should be able to refactor it one day -- so all that matters now is that outward interface is ~ok. If you want to give a try to the `from_str` without the extra parse_filename_as but still supporting it, that would be cool. But I'm fine with the current redundant state just to be safe, and simplifying later (we would then make the param Optional, and if specified, prefer it, to preserve compatibility)

Reply:
How would you implement it? When you have sth like `parser = AutoParser.from_str("col=b/fname")`, does `fname` stands for filename or does it mean folder with an implicit `*` value like [in this example](https://github.com/AmpX-AI/fsql/blob/main/tests/test_example_usage.py#L76)? I guess that's what you are reffering to as "a risky business". You would basically need to auto-parse something like `parser = AutoParser.from_str("col=b/colX/fname")`. We could do sth like global filename partition name (some smarter version of `FILENAME_PARTITION = "fname"`) and if present in path or allowed, add it. Otherwise I do not see a simple way how to distinguish it (we would probably need to do it on the fly, but that would probably require some changes in the partition discovery).

Reply:
yeah, thats exactly one of the pitfalls. The global filename does not look good to me, instead, we could eg:
- add a second param -- `AutoParser.from_str("col1=b/col2", "my_fname")` (and you could eg `"my_fname=[f1.parquet, f2.parquet]"`). We would do it also for FixedParser, and yell `deprecated` when the original behaviour would be relied upon
- invalidate this example -- if ends with `/`, then its a directory, otherwise its a filename. Apply in both Auto and Fixed parsers. Backwards compatibility requires we make `from_str_v2` instead

in either approach, the next major version would simplify code and remove the deprecated part / extraneous code. I'm leaning towards the first one, it is less hacky, less error-prone in some way (trailing `/` is a menace), albeit more verbose and more error-prone in another way (you do not realise you should have used it and still try to squeeze fname to the single string).

Reply:
sorry for a delay 😇 

I am still unsure about the right solution. I would like to have a consistent interface for both parsers, but I still think it would need a major refactoring (and it's even more confusing as there is the `DateRangeGenerator` "parser" which shares the base column parser as well). I definitely do not want to go with a `/` (as you wrote, trailing `/` is a menace). It seems good to me to separate path description and filename (which is basically needed for AutoParser to correctly parse and query filenames as well). But the current solution is still not an ideal imo.

Reply:
I kept thinking about it, and I believe now that there is exactly one way out -- a genuine grammar. I've recently did a grammar to translate sql queries to pandas sql expressions in https://github.com/erikrose/parsimonious/tree/master and it turned out quite good. It will bring us here a few advantages -- in particular, separating-decoupling between users specifying what they want (`from_str`, generating partitions, ...) and how we interpret it (auto parses / fixed columns parser). As well as independent testability and better catching of bad inputs and providing expl messages

that will be a rather big refactoring, with incompatible interfaces. We'll essentially have to create `ColumnParserV2(ABC)`, with a different from_str, and accept either union or a protocol inside the impl. That makes me fully ok with your 

> I am afraid there is no simple solution

comment, which I agree with. Let's merge this, in particular the tests -- those will be the most useful thing during refactoring.

Reply:
Interesting, was the grammer you did something proprietary or is it shared somewhere? It really sounds it could help us here a lot, but as you wrote it would be a big refactoring.

Reply:
that is proprietary atm, it is full of business-specific tweaks and not general enough. And I may not even attempt to make it publishable -- I'll first need to check whether https://github.com/tobymao/sqlglot doesn't always do the same thing but better... Either way, I'd have to write a wholly new grammar here. Ideally, it would also cover types to some extent (and you would not have to distinguish whether comparisons should be numerical or string), etc etc...